### PR TITLE
Include the parent of iframe element when searching for closest selector

### PR DESCRIPTION
**Pull request by avimehta**
[WIP]: Need to write tests for dom.js and visibility-impl.js changes.

For #5657

### CODE REVIEW

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 1]
```diff
@@ -160,6 +161,10 @@ export function getElement(selector, el, selectionMethod) {
     return closestByTag(el, selector);
   } else if (selectionMethod == 'scope') {
     return el.parentElement.querySelector(selector);
+  } else if (selectionMethod == 'host') {
```
Comment (jridgewell) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Do we need to test that this is a friendly iframe?

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 2]
```diff
@@ -160,6 +161,10 @@ export function getElement(selector, el, selectionMethod) {
...
+    const elWin = el.ownerDocument.defaultView;
+    const parentEl = elWin.frameElement && elWin.frameElement.parentElement;
+    return closestBySelector(parentEl, '.-amp-element');
```
Comment (jridgewell) [inline extensions/amp-analytics/0.1/visibility-impl.js]: This has to be a `!Element`, right? But it could be null from the `elWin.frameElement` test.

File: src/dom.js [hunk 3]
```diff
@@ -204,6 +204,49 @@ export function closestByTag(element, tagName) {
...
+        el.msMatchesSelector ||
+        el.oMatchesSelector;
+    if (matcher) {
+      return matcher.call(el, selector);
+    }
+    if (opt_callback) {
+      return opt_callback(el, selector);
+    }
+  } catch (selectorError) {
+    user().error('DOM', 'Bad query selector.', selector, selectorError);
```
Comment (jridgewell) [inline src/dom.js]: Seems like a dev error.

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 2 above)
Comment (avimehta) [inline extensions/amp-analytics/0.1/visibility-impl.js]: good point. Will add a check for this.

File: src/dom.js (hunk 3 above)
Comment (avimehta) [inline src/dom.js]: selector comes from user (via instrumentation.js) and from dev (via visibility-impl.). I don't expect visibility-impl to change hence went with user error.

Comment (jridgewell) [inline src/dom.js]: Ah, 👍

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 1 above)
Comment (jridgewell) [inline extensions/amp-analytics/0.1/visibility-impl.js]: 👍

Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: So, I've looked around some more. I still think it's very conventional to support support `sectionMethod == 'host' || selector = ':host' || selector = ':root'`.

But the bigger question I have, from embed's point of view, what does this trigger mean:
```
"triggers": {
  "totalVisibility": {
    "on": "visible",
    "request": "visibility",
  }
}
```

Would we actually allow an embed to get the analytics event for page visibility? It seems misleading, since we never impress embeds when page is NOT visible.

I think in the scope of the embed's document, the most logical meaning of this is that the embed wants to known its own visibility. Thus, I believe that `!visibilitySpec` and `selectorMethod = host || selector = :host` are actually two different ways to ask for page-level visibility event and it's simply that interpretation of this is different whether the `amp-analytics` is in the main document or embed. This is not at all far-fetched - this is basically the same interpretation as we'd have with `window.onload` and `iframe.contentWindow.onload`.

So, the actual code structure I see:
```
if (!visibilitySpec || visibilitySpec.selectorMethod = 'host' || ...:host) {
  // This is either page-level visibility or embed visibility.
  if (isEmbed(contextElement)) {
    ...
  } else {
    viewerFor(...).whenVisible().then(() => {
      trigger();
    });
  }
} else {
  // This is element visibility - no change from current implementation.
}
```

File: src/dom.js (hunk 3 above)
Comment (dvoytenko) [inline src/dom.js]: This should not be a `user` error. We shouldn't catch the error here at all. The caller should make that determination. E.g. `closestSelector('.-amp-element')` is not a user-entered selector and thus cannot cause a user error. But where we parse selectors from user config - those are user error. 

Bottom line; let's remove try/catch from here and restore try/catch/user-error  here: https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/0.1/instrumentation.js#L482

**Comment by avimehta** [comment]
This PR is going to change a bit. Skip review for now.

**Comment by avimehta** [comment]
Ready for review again.

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 4]
```diff
@@ -102,9 +102,10 @@ export function isVisibilitySpecValid(config) {
 
   const spec = config['visibilitySpec'];
   const selector = spec['selector'];
-  if (!selector || (selector[0] != '#' && selector.indexOf('amp-') != 0)) {
-    user().error(TAG_, 'Visibility spec requires an id selector or a tag ' +
-        'name starting with "amp-"');
+  if (!selector || (selector[0] != '#' && selector.indexOf('amp-') != 0 &&
```
Comment (jridgewell) [inline extensions/amp-analytics/0.1/visibility-impl.js]: `startsWith`.

File: src/dom.js (hunk 3 above)
Comment (jridgewell) [inline src/dom.js]: I agree with @dvoytenko.

File: src/dom.js [hunk 5]
```diff
@@ -204,6 +204,49 @@ export function closestByTag(element, tagName) {
...
+ * @param {string} selector
+ * @return {?Element} closest ancestor if found.
+ */
+export function closestBySelector(element, selector) {
+  if (element.closest) {
+    return element.closest(selector);
+  }
+
+  return closest(element, el => {
+    return matches(el, selector);
```
Comment (jridgewell) [inline src/dom.js]: Do we need a fallback implementation?

File: src/dom.js [hunk 6]
```diff
@@ -204,6 +204,49 @@ export function closestByTag(element, tagName) {
...
+  });
+}
+
+/**
+ * Checks if the given element matches the selector
+ * @param  {!Element} el The element to verify
+ * @param  {!string} selector The selector to check against
+ * @param  {!function(!Element, string):boolean=} opt_callback Function to call
```
Comment (jridgewell) [inline src/dom.js]: Let's call this `opt_fallback`

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 4 above)
Comment (avimehta) [inline extensions/amp-analytics/0.1/visibility-impl.js]: [11:53:47] Not available because we do not currently ship with a needed ES6 polyfill. Is `string.startsWith` not the right form?

File: src/dom.js (hunk 5 above)
Comment (avimehta) [inline src/dom.js]: It is used in instrumentation.js. Happy to go with an alternate route but can't think of anything clean. There are two reasons for the callback:

1) instrumentation.js has a different matching behaviour then closestBySelector. 

2) `matches()` will return false for both "`.matches` method doesn't exist" and "doesn't match". As a result, if we act on the return value, we would do a second round of matching even when it is not needed(`.matches` exists and the element did not match the selector).

File: src/dom.js (hunk 3 above)
Comment (avimehta) [inline src/dom.js]: done.

File: src/dom.js (hunk 6 above)
Comment (avimehta) [inline src/dom.js]: done.

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 7]
```diff
@@ -162,7 +162,11 @@ export function getElement(selector, el, selectionMethod) {
     const elWin = el.ownerDocument.defaultView;
     const parentEl = elWin.frameElement && elWin.frameElement.parentElement;
     if (parentEl) {
-      return closestBySelector(parentEl, '.-amp-element');
+      try {
+        return closestBySelector(parentEl, '.-amp-element');
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/visibility-impl.js]: done.

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 4 above)
Comment (avimehta) [inline extensions/amp-analytics/0.1/visibility-impl.js]: found the polyfill in src/strings. Using that.

**Comment by avimehta** [comment]
ptal.

File: src/dom.js (hunk 5 above)
Comment (jridgewell) [inline src/dom.js]: I'm not sure we're talking about the same thing? I mean do we need to code a fallback cb if `matches` is not supported natively.

Comment (jridgewell) [inline src/dom.js]: 🔪, IE8 is dead. Let's remove all the fallbacks. 🎉

File: extensions/amp-analytics/0.1/instrumentation.js [hunk 8]
```diff
@@ -466,18 +466,12 @@ export class InstrumentationService {
...
-          el.oMatchesSelector;
-      if (matcher) {
-        return matcher.call(el, selector);
-      }
-      const matches = el.ownerDocument.querySelectorAll(selector);
-      let i = matches.length;
-      while (i-- > 0 && matches.item(i) != el) {};
-      return i > -1;
+      return matches(el, selector, () => {
+        const matches = el.ownerDocument.querySelectorAll(selector);
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/instrumentation.js]: Shouldn't this part simply go inside `matches` utility? This is a polyfill, right? For a rare case of browsers that still need it, this seems totally justifiable.

File: extensions/amp-analytics/0.1/test/test-visibility-impl.js [hunk 9]
```diff
@@ -412,5 +418,10 @@ describe('amp-analytics.visibility', () => {
       expect(getElement('div', analytics, 'scope')).to.equal(null);
       expect(getElement('amp-img', analytics, 'scope')).to.equal(img2);
     });
+
+    it.skip('finds element for selectionMethod=host', () => {
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/test/test-visibility-impl.js]: To initialize iframe fully what you need to do is this:
```
beforeEach(() => {
  const loaded = loadPromise(iframe);
  container.appendChild(iframe);
  return loaded;
});
```

See an example here: https://github.com/ampproject/amphtml/blob/master/test/functional/test-extensions.js#L398

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 10]
```diff
@@ -155,6 +159,16 @@ export function getElement(selector, el, selectionMethod) {
   if (!el) {
     return null;
   }
+
+  // Special case for root selector.
+  if (selector == ':host' || selector == ':root') {
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: So this case now should also support `!selector`, right?

File: extensions/amp-analytics/0.1/instrumentation.js (hunk 8 above)
Comment (avimehta) [inline extensions/amp-analytics/0.1/instrumentation.js]: done.

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 10 above)
Comment (avimehta) [inline extensions/amp-analytics/0.1/visibility-impl.js]: I would prefer to do this in another PR. Currently, if there is no `visibilitySpec`, this method is never called. If we want to default selector to something, instrumentation.js will need to be updated to either call this always or add the selector there for analytics inside embeds.

**Comment by avimehta** [comment]
Ptal

File: src/dom.js (hunk 5 above)
Comment (dvoytenko) [inline src/dom.js]: My point is this: if we decide we do NOT need a fallback, let's eliminate completely. If we decided that we DO need fallback - it should be inside the `matches` method. So, let's confirm for sure. If this is ONLY b/c of IE8 - let's remove the fallback.

Comment (avimehta) [inline src/dom.js]: ok.. removed again :)

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 10 above)
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Ok. Please file a bug then for yourself.

**Comment by dvoytenko** [comment]
LGTM with a request to file an issue for `visibilitySpec` redefinition.

**Comment by avimehta** [comment]
Filed: https://github.com/ampproject/amphtml/issues/6128
